internal/util/util.go

```diff
@@ -322,39 +324,106 @@ func ParseClientIP(addr string) (string, error) {
+//   - ErrImageNotFound: The image doesn't exist in the rbd pool.
+//   - rados.ErrPermissionDenied: Permissions to access the pool is denied.
+//
+// If any of these errors are encountered, the function returns `true`, indicating
+// that the volume search should continue because of known error. Otherwise, it
+// returns `false`, meaning the search should stop.
+//
+// This helper function is used in scenarios where multiple attempts may be made
+// to retrieve or generate RBD volume information, and we want to gracefully handle
+// specific failure cases while retrying for others.
+func ShouldRetryRBDVolumeGeneration(err error) bool {
+	if err == nil {
+		return false // No error, do not retry
+	}
+	// Continue searching for specific known errors
+	return (errors.Is(err, ErrKeyNotFound) ||
+		errors.Is(err, ErrPoolNotFound) ||
+		errors.Is(err, ErrConfigNotFound) ||
+		errors.Is(err, rbderrors.ErrImageNotFound) ||
+		errors.Is(err, rados.ErrPermissionDenied))
+}
+
 // GetControllerPublishSecretRef retrieves the controller publish secret from ceph-csi-config ConfigMap
 // for a given clusterID. Fetches the secret from Kubernetes, and returns it as a map of key-value pairs.
 func GetControllerPublishSecretRef(volumeId, driverType string) (string, string, error) {
 	var (
-		vi               CSIIdentifier
-		secretName       string
-		secretNamespace  string
-		getSecretRefFunc func(string, string) (string, string, error)
+		vi              CSIIdentifier
+		secretName      string
+		secretNamespace string
 	)
 	err := vi.DecomposeCSIID(volumeId)
 	if err != nil {
 		return secretName, secretNamespace, fmt.Errorf("failed to decode volume ID (%s): %w", volumeId, err)
 	}
 
-	switch driverType {
-	case RBDType:
-		getSecretRefFunc = GetRBDControllerPublishSecretRef
-	case CephFsType:
-		getSecretRefFunc = GetCephFSControllerPublishSecretRef
-	default:
-		return secretName, secretNamespace, fmt.Errorf("unsupported driver type: %s", driverType)
-	}
-
-	secretName, secretNamespace, err = getSecretRefFunc(CsiConfigFile, vi.ClusterID)
-	if err != nil {
+	secretName, secretNamespace, err = getControllerPublishSecretRef(vi.ClusterID, driverType)
+	if !ShouldRetryVolumeGeneration(err) {
 		return secretName, secretNamespace,
 			fmt.Errorf("failed to get controller publish secret details from csi config file: %w", err)
 	}
 
+	if secretName != "" && secretNamespace != "" {
+		return secretName, secretNamespace, nil
+	}
+
+	// Check clusterID mapping exists
+	mapping, mErr := GetClusterMappingInfo(vi.ClusterID)
+	if mErr != nil {
+		return secretName, secretNamespace, mErr
+	}
+	if mapping != nil {
+		for _, cm := range *mapping {
+			for key, val := range cm.ClusterIDMapping {
+				mappedClusterID := GetMappedID(key, val, vi.ClusterID)
+				if mappedClusterID == "" {
+					continue
+				}
+
+				secretName, secretNamespace, err := getControllerPublishSecretRef(mappedClusterID, driverType)
+				if !ShouldRetryVolumeGeneration(err) {
```

Comment:
why do we need this change because we know that we will get only `ErrConfigNotFound` no other RBD specific error here. lets not use `ShouldRetryVolumeGeneration` here which seems redundant